### ISSUE TITLE
Bug: Server hydration mistmatch and radio group with defaultChecked

### ISSUE THREAD

**Issue by leoasis** [facebook/react#18502]
When hydrating from the server, whenever there's a mismatch in the initially checked input in a radio button group and the inputs use `defaultChecked` (uncontrolled), there is no warning of that mismatch and the component behaves in a buggy way for the input that was initially selected in the server payload (the `onChange` callback prop is not fired for the input that was marked as selected in the initial html, when selecting it).

Worth noting that using a controlled input (using `checked`) makes the bug go away.

React version: 16.13.1

## Steps To Reproduce

1. In the example provided below, make sure you refresh the browser within Code Sandbox.
2. Check the first radio button. Verify that the radio is checked, but the text next to it still shows as "not checked" (which means the onChange prop was not triggered)
3. Attempt to select any other option (works fine), and then back the first one (works fine too).
4. Refreshing the page again, choosing any other option other than the first works fine.

Link to code example: https://codesandbox.io/s/affectionate-stonebraker-3wj68

## The current behavior

I know that using hydrate doesn't guarantee fixing the mismatches, but it says that it should warn about them in development (https://reactjs.org/docs/react-dom.html#hydrate). In this case, React doesn't warn about the mismatch, and the behavior is buggy, as shown in the example and steps to repro.

## The expected behavior

React should ideally work without the issue, but given that it's not guaranteed to be fixed according to the documentation, it should at least warn about it.

**Comment by gaearon** [comment]
Interested in looking into a fix?

**Comment by leoasis** [comment]
Yeah! Would the fix in this case be to show the mismatch warning or to make onChange work fine for the input initially checked?

**Comment by gaearon** [comment]
We should warn about mismatch and do whatever's simplest. (because when you have a mismatch, we don't give any guarantees)

**Comment by HeroProtagonist** [comment]
👋 
Started to take a look at this one and have some findings. 

[`shouldIgnoreAttribute`](https://github.com/facebook/react/blob/b93f3e7d2d595444c6ee2964d4cce2e17f4ffa28/packages/react-dom/src/client/ReactDOMComponent.js#L1142) returns `true` for `defaultChecked` because `propertyInfo.type === RESERVED`. So any warning in that code block is skipped.

```js
...
      } else if (
        !shouldIgnoreAttribute(propKey, propertyInfo, isCustomComponentTag) &&
...
```

Also `checked` attribute seems to be entirely [skipped](https://github.com/facebook/react/blob/b93f3e7d2d595444c6ee2964d4cce2e17f4ffa28/packages/react-dom/src/client/ReactDOMComponent.js#L1109) over when looking for mismatches.

```js
...
      } else if (
        propKey === SUPPRESS_CONTENT_EDITABLE_WARNING ||
        propKey === SUPPRESS_HYDRATION_WARNING ||
        // Controlled attributes are not validated
        // TODO: Only ignore them on controlled tags.
        propKey === 'value' ||
        propKey === 'checked' ||
        propKey === 'selected'
      ) {
        // Noop
      } 
...
``` 

I would be happy to submit a PR, but would need some guidance on how to proceed. Thanks

**Comment by gaearon** [comment]
Does this repro with React 18? I tried to re-run the sandbox but HTML is inlined so it's hard to say what exactly was supposed to mismatch. The example as is does error in 18.
